clash-prelude/src/Clash/Class/Finite/Internal.hs

```diff
@@ -0,0 +1,1006 @@
+    SNatLE -> const lowest
+    SNatGT -> case zeroLeAdd @(ElementCount b) @1 of
+      Dict -> case powPositiveImpliesPositiveBase
+                     @(ElementCount b) @(ElementCount a) of {}
+
+  lowestMaybe  = const <$> lowestMaybe
+
+  highest = case compareSNat (SNat @1) (SNat @(ElementCount b)) of
+    SNatLE -> const highest
+    SNatGT -> case zeroLeAdd @(ElementCount b) @1 of
+      Dict -> case powPositiveImpliesPositiveBase
+                     @(ElementCount b) @(ElementCount a) of {}
+
+  highestMaybe = const <$> highestMaybe
+
+  predMaybe f = do
+    h <- highestMaybe
+    either Just (const Nothing)
+      $ foldr (\i -> (=<<) $ \g -> do
+                 let g' y x = if index x == i then y else g x
+                 maybe (Right $ g' h) (Left . g') $ predMaybe $ g $ ith i
+              ) (Right f) $ indicesI @(ElementCount a)
+
+  succMaybe f = do
+    l <- lowestMaybe
+    either Just (const Nothing)
+      $ foldr (\i -> (=<<) $ \g -> do
+                 let g' y x = if index x == i then y else g x
+                 maybe (Right $ g' l) (Left . g') $ succMaybe $ g $ ith i
+              ) (Right f) $ indicesI @(ElementCount a)
+
+  ith      = ((. index) . (!!)) . ith @(Vec (ElementCount a) b)
+  index f  = index (f . ith <$> indicesI)
+
+instance a ~ b => Finite (a :~: b) where
+  type ElementCount (a :~: b) = 1
+  elements     = Refl :> Nil
+  lowest       = Refl
+  lowestMaybe  = Just Refl
+  highest      = Refl
+  highestMaybe = Just Refl
+  predMaybe    = const Nothing
+  succMaybe    = const Nothing
+  ith          = const Refl
+  index        = const 0
+
+instance a ~~ b => Finite (a :~~: b) where
+  type ElementCount (a :~~: b) = 1
+  elements     = HRefl :> Nil
+  lowest       = HRefl
+  lowestMaybe  = Just HRefl
+  highest      = HRefl
+  highestMaybe = Just HRefl
+  predMaybe    = const Nothing
+  succMaybe    = const Nothing
+  ith          = const HRefl
+  index        = const 0
+
+#if MIN_VERSION_base(4,15,0)
+-- | Reverses the index order used by the 'Finite' instance of the inner type.
+--
+-- >>> elements @(Maybe Bool)
+-- Nothing :> Just False :> Just True :> Nil
+--
+-- >>> elements @(Down (Maybe Bool))
+-- Down (Just True) :> Down (Just False) :> Down Nothing :> Nil
+--
+instance Finite a => Finite (Down a) where
+  type ElementCount (Down a) = ElementCount a
+  elements     = Down <$> reverse elements
+  lowest       = Down highest
+  lowestMaybe  = Down <$> highestMaybe
+  highest      = Down lowest
+  highestMaybe = Down <$> lowestMaybe
+  predMaybe    = fmap Down . succMaybe . getDown
+  succMaybe    = fmap Down . predMaybe . getDown
+  ith          = Down . ith . (maxBound -)
+  index        = (maxBound -) . index . getDown
+#endif
+
+-- | A newtype wrapper that reverses the index order, which normally
+-- would be used by the 'Finite' instance of the inner type via
+-- 'Generic' deriving. The newtype is only intended be used with the
+#if MIN_VERSION_base(4,15,0)
```

Comment:
This does not appear to work. CPP and Haddock do not play nice. I think it works better with the `{-` construction, but still, you'll have to jump through some hoops and you're probably better off with a little duplication. Otherwise you'll get CPP-generated markers as part of your Haddock.

Please test with GCC versions both matching the `#if` and the `#else`.

Note how your documentation ends at "used with the" in the Haddock generated in CI:
https://clash-lang.gitlab.io/-/clash-compiler/-/jobs/8839034866/artifacts/hadocs/clash-prelude/clash-prelude/Clash-Class-Finite.html#t:GenericReverse